### PR TITLE
Address CRC incompatibility by moving to protocol version 2

### PR DESCRIPTION
**Pull request by kellymclaughlin**
This PR presents an alternative solution to the problem of CRC incompatibility that was originally discussed in #23. Changes to address #23 were part of [this](https://github.com/joyent/node-fast/issues/23) commit. Subsequent to the merge of those changes a problem was discovered. It was assumed that the CRC calculated by the outdated version (0.3.0) of node-crc was always incorrect, but there are cases where the CRC code  calculated is correct. This makes it impossible to deduce which version of node-crc was used to encode a received message and therefore there is no way to know which version should  be used to encode a response. PR #27 presents one solution to this problem, but it has some undesirable worst case performance consequences.

This PR takes a different approach to solve the CRC incompatibility problem. Instead of trying to transparently determine which version of the CRC library should be used to decode a received message and encode a response this change updates the protocol to version 2. Version 2 explicitly uses the updated version of the CRC library for both fast servers and clients. Fast servers using version 2 are able to communicate with version 2 clients as well as clients still using version 1 of the protocol (*i.e.* clients using the outdated version of node-crc). However, clients using version 2 are not able to work with servers using protocol version 1. This is the only restriction to this solution and the trade-off is a very predictable means to determine which version of node-crc to use and no potential performance penalty as is the case with #27.

### CODE REVIEW

File: lib/fast_client.js [hunk 1]
```diff
@@ -75,12 +75,6 @@ fastNclients = 0;
  *     			request failures.  The caller is also expected to listen
  *     			for these errors and handle reconnection appropriately.
  *
- *     crc_mode		CRC mode. Used to transition off buggy node-crc version.
- *             		See the fast_protocol module documentation for more
- *                      details. The valid values for a FastClient are
- *                      FAST_CHECKSUM_V1 and FAST_CHECKSUM_V2.
- *                      FAST_CHECKSUM_V1_V2 is only valid for FastServers.
- *
  * On 'error', the caller should assume that the current connection to the
```
Comment (trentm) [inline lib/fast_client.js]: Should there be a comment addition for the new "version" option?

File: lib/fast_protocol.js [hunk 2]
```diff
@@ -223,105 +212,58 @@ function fastMessageEncode(msg, default_crc_mode)
...
- *                      May be undefined if the CRC mode is set to
- *                      FAST_CHECKSUM_V2.
- *     v2CalculatedCrc  The calculated CRC for the FAST_CHECKSUM_V2 CRC mode.
- *                      May be undefined if the CRC mode is set to
- *                      FAST_CHECKSUM_V1.
+ *     calculatedCrc    The calculated CRC for the received message.
  */
-function crcValidationError(crcMode, headerCrc, v1CalculatedCrc,
-    v2CalculatedCrc) {
+function crcValidationError(headerCrc, calculatedCrc) {
```
Comment (trentm) [inline lib/fast_protocol.js]: I've never looking into debugging when CRC errors show up in INC/SUP tickets. Would it help if the protocol version was in the error message?

File: lib/fast_protocol.js [hunk 3]
```diff
@@ -223,105 +212,58 @@ function fastMessageEncode(msg, default_crc_mode)
...
-		errorObj.info.crcCalculated = v2CalculatedCrc;
-		crcStr = mod_extsprintf.sprintf('%s', v2CalculatedCrc);
-		break;
-	default:
-		break;
-	}
+	    '%s, found %s', headerCrc, calculatedCrc);
 
-	errorStr += crcStr;
 	return (new VError(errorObj, errorStr));
```
Comment (trentm) [inline lib/fast_protocol.js]: nit: VError does the sprintf thing already, so this *could* be reduced to:

```
return (new VError(errorObj, 'fast protocol: expected CRC %s, found %s', headerCrc, calculatedCrc));
```

File: bin/fastserve [hunk 4]
```diff
@@ -32,8 +32,6 @@ var OPTS = {
     '-p,--port': 'fast RPC listening port(default: 2030)',
     '-m,--mon-port': 'monitoring server listening port (default: port+800)',
     '-q,--quiesce': 'enable quiesce (default: false)',
```
Comment (trentm) [inline bin/fastserve]:
```suggestion
    '-q,--quiesce': 'enable quiesce (default: false)'
```

for 'make check'

Comment (trentm) [reply bin/fastserve]: I don't grok why the UI here is showing that I changed the leading whitespace. I don't think I did.

File: lib/fast_server.js [hunk 5]
```diff
@@ -465,7 +449,6 @@ FastServer.prototype.connCreate = function (sock)
 	    'connId': cid,
 	    'socket': sock,
 	    'log': this.fs_log,
```
Comment (trentm) [inline lib/fast_server.js]:
```suggestion
	    'log': this.fs_log
```

for `make check`

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,1280 @@
+{
+	"name": "fast",
+	"version": "2.8.1",
```
Comment (trentm) [inline package-lock.json]: This doesn't have version 3.0.0.

File: lib/fast_server.js [hunk 7]
```diff
@@ -174,11 +174,6 @@ fastNservers = 0;
  *
```
Comment (trentm) [inline lib/fast_server.js]: nit: Copyright year above is  not updated.

File: test/tst.client_generic.js [hunk 8]
```diff
@@ -506,14 +509,8 @@ var test_cases = [ {
...
-			var cause = rq.ctr_error.cause();
-			mod_assertplus.equal(cause.name, 'FastTransportError');
-			mod_assertplus.ok(
-			    /^unexpected error on transport:/.test(
-			    cause.message));
-
-			cause = cause.cause();
-			mod_assertplus.ok(/ECONNRESET/.test(cause.message));
+			var name = rq.ctr_error.cause().name;
+			mod_assertplus.ok(name === 'FastTransportError' || name === 'FastProtocolError');
```
Comment (trentm) [inline test/tst.client_generic.js]: Is there where you are cheating? ;)

File: test/tst.protocol_encoder.js [hunk 9]
```diff
@@ -255,8 +298,8 @@ function runTestCase(testcase)
 		 */
 		mod_assertplus.ok(Buffer.isBuffer(outbuf));
 		mod_assertplus.ok(outbuf.length > mod_protocol.FP_HEADER_SZ);
-		mod_assertplus.equal(1,
-		    outbuf.readUInt8(mod_protocol.FP_OFF_VERSION));
+//		mod_assertplus.equal(mod_protocol.FP_VERSION_CURRENT,
+//		    outbuf.readUInt8(mod_protocol.FP_OFF_VERSION));
```
Comment (trentm) [inline test/tst.protocol_encoder.js]: Forgotten debugging changes here?

File: test/tst.client_generic.js (hunk 8 above)
Comment (kellymclaughlin) [reply test/tst.client_generic.js]: Hehehe, yep that's it!

File: test/tst.protocol_encoder.js (hunk 9 above)
Comment (kellymclaughlin) [reply test/tst.protocol_encoder.js]: Oh yeah, those just need to be deleted. The version checks moved into the individual test cases now.

File: lib/fast_protocol.js (hunk 2 above)
Comment (kellymclaughlin) [reply lib/fast_protocol.js]: Good idea, will do.

File: package-lock.json (hunk 6 above)
Comment (kellymclaughlin) [reply package-lock.json]: I'm not sure of the correct way to update this file. What's the recommended procedure?

Comment (trentm) [reply package-lock.json]: Re-running `npm install` with a new-enough npm should do it.  TBH I'm not totally comfortable on package-lock.json maintenance myself... especially when we commonly have so much old npm version usage still in play in various repos.

**Pull request merged**